license_manager/service/service.go

```diff
@@ -2,36 +2,295 @@ package service
 
 import (
 	"context"
+	"sync"
+	"time"
 
 	lmpb "github.com/enfabrica/enkit/license_manager/proto"
 
+	"github.com/google/uuid"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/timestamppb"
 )
 
-type Service struct{}
+// Service implements the LicenseManager gRPC service.
+type Service struct {
+	mu           sync.Mutex          // Protects the following members from concurrent access
+	currentState state               // State of the server
+	licenses     map[string]*license // Queues and allocations, managed per-license-type
 
+	queueRefreshDuration      time.Duration // Queue entries not refreshed within this duration are expired
+	allocationRefreshDuration time.Duration // Allocations not refreshed within this duration are expired
+}
+
+func New() *Service {
+	service := &Service{
+		currentState: stateStarting,
+		// TODO: Read this from a config file and pass it in
+		licenses: map[string]*license{
+			"xilinx::foo": &license{
+				totalAvailable: 2,
+				allocations:    map[string]*invocation{},
+			},
+			"xilinx::bar": &license{
+				totalAvailable: 2,
+				allocations:    map[string]*invocation{},
+			},
+		},
+		queueRefreshDuration:      5 * time.Second,
+		allocationRefreshDuration: 5 * time.Second,
+	}
+
+	go func(s *Service) {
+		t := time.NewTicker(1 * time.Second)
```

Comment:
`1 * time.Second` should probably also come from flag.

Reply:
Added TODO